beacon/goclient/attest.go

```diff
@@ -27,22 +28,73 @@ func (gc *GoClient) AttesterDuties(ctx context.Context, epoch phase0.Epoch, vali
 	return resp.Data, nil
 }
 
-func (gc *GoClient) GetAttestationData(slot phase0.Slot, committeeIndex phase0.CommitteeIndex) (*phase0.AttestationData, spec.DataVersion, error) {
-	attDataReqStart := time.Now()
-	resp, err := gc.client.AttestationData(gc.ctx, &api.AttestationDataOpts{
-		Slot:           slot,
-		CommitteeIndex: committeeIndex,
-	})
-	if err != nil {
-		return nil, DataVersionNil, fmt.Errorf("failed to get attestation data: %w", err)
+// GetAttestationData returns attestation data for a given slot (which is same for all 64 committeeIndex
+// values that identify Ethereum committees chosen to attest on this slot).
+// Note, committeeIndex is an optional parameter that will be used to set AttestationData.Index
+// in the resulting data returned from this function.
+// Note, result returned is meant to be read-only, it's not safe to modify it (because it will be
+// accessed by multiple concurrent readers).
+func (gc *GoClient) GetAttestationData(slot phase0.Slot, committeeIndex phase0.CommitteeIndex) (
+	*phase0.AttestationData,
+	spec.DataVersion,
+	error,
+) {
+	// Check cache.
+	cachedResult := gc.attestationDataCache.Get(slot)
+	if cachedResult != nil {
+		return withCommitteeIndex(cachedResult.Value(), committeeIndex), spec.DataVersionPhase0, nil
 	}
-	if resp == nil {
-		return nil, DataVersionNil, fmt.Errorf("attestation data response is nil")
+
+	// Have to make beacon node request and cache the result.
+	result, err := func() (*phase0.AttestationData, error) {
+		// Requests with the same slot number must lock on the same mutex.
+		reqMu := &gc.attestationReqMuPool[uint64(slot)%uint64(len(gc.attestationReqMuPool))]
+		reqMu.Lock()
+		defer reqMu.Unlock()
+
+		// Prevent making more than 1 beacon node requests in case somebody has already made this
+		// request concurrently and succeeded.
+		cachedResult := gc.attestationDataCache.Get(slot)
+		if cachedResult != nil {
+			return cachedResult.Value(), nil
+		}
+
+		attDataReqStart := time.Now()
+		resp, err := gc.client.AttestationData(gc.ctx, &api.AttestationDataOpts{
+			Slot: slot,
+		})
+		metricsAttesterDataRequest.Observe(time.Since(attDataReqStart).Seconds())
+		if err != nil {
+			return nil, fmt.Errorf("failed to get attestation data: %w", err)
+		}
+		if resp == nil {
+			return nil, fmt.Errorf("attestation data response is nil")
+		}
+
+		gc.attestationDataCache.Set(slot, resp.Data, ttlcache.DefaultTTL)
+
+		return resp.Data, nil
+	}()
+	if err != nil {
+		return nil, DataVersionNil, err
 	}
```

Comment:
this should work and seems to address the various edgecases 👌 

with that said, i think this could be massively simplified down to just one struct field and one function call, using https://pkg.go.dev/golang.org/x/sync/singleflight

while it's not ideal that singleflight isn't using generics, it should still be much easier to reason about the code  than the advanced implementation we have currently (imo)

Reply:
I was looking for something like this ( turns out not hard enough :) ), thanks for bringing it up!

Although I was trying to find a package that handles both aspects: **inflight** + **caching**. Still, https://pkg.go.dev/tailscale.com/util/singleflight does simplify this code a bit (and unit tests don't complain) - see https://github.com/ssvlabs/ssv/pull/1798/commits/fcaa83d2825c1cc2fc7bd8d196d82b65e705b249 for how it's different compared to the code I had before.

From performance perspective, I don't think it matters much for this particular case, but if we were to push it to the limits - I think each of these 2 approaches will have it's preferred type of workload while doing worse in some circumstances.

Reply:
@iurii-ssv

+1

the readability gain is worth the performance cost here IMO

(don't think performance cost is noticeable below many thousands of calls per second)